Zend/zend_language_parser.y

```diff
@@ -1170,18 +1172,39 @@ property_name:
 ;
 
 assignment_list:
-		assignment_list ',' assignment_list_element
+		unkeyed_assignment_list
+			{ $$ = $1; }
+	|	keyed_assignment_list possible_comma
+			{ $$ = $1; }
+;
+
+unkeyed_assignment_list:
+		unkeyed_assignment_list ',' unkeyed_assignment_list_element
 			{ $$ = zend_ast_list_add($1, $3); }
-	|	assignment_list_element
+	|	unkeyed_assignment_list_element
 			{ $$ = zend_ast_create_list(1, ZEND_AST_LIST, $1); }
 ;
 
-assignment_list_element:
+unkeyed_assignment_list_element:
 		variable						{ $$ = $1; }
 	|	T_LIST '(' assignment_list ')'	{ $$ = $3; }
 	|	/* empty */						{ $$ = NULL; }
 ;
 
+keyed_assignment_list:
+		keyed_assignment_list ',' keyed_assignment_list_element
+			{ $$ = zend_ast_list_add($1, $3); }
+	|	keyed_assignment_list_element
+			{ $$ = zend_ast_create_list(1, ZEND_AST_LIST, $1); }
+;
+
+keyed_assignment_list_element:
+		expr T_DOUBLE_ARROW variable
+			{ $$ = zend_ast_create(ZEND_AST_ARRAY_ELEM, $3, $1); }
+	|	expr T_DOUBLE_ARROW T_LIST '(' assignment_list ')'
+			{ $$ = zend_ast_create(ZEND_AST_ARRAY_ELEM, $5, $1); }
+;
```

Comment:
That would be necessary if we ever replace `list()` with `[]`, but the way I've done it thus far makes things neater.